x-pack/plugins/lens/public/indexpattern_plugin/operations.ts

```diff
@@ -5,157 +5,104 @@
  */
 
 import _ from 'lodash';
-import { i18n } from '@kbn/i18n';
-import { DataType, DimensionPriority } from '../types';
+import { DimensionPriority } from '../types';
 import {
   IndexPatternColumn,
   IndexPatternField,
   IndexPatternPrivateState,
   OperationType,
+  BaseIndexPatternColumn,
 } from './indexpattern';
+import { termsOperation } from './operation_definitions/terms';
+import {
+  minOperation,
+  averageOperation,
+  sumOperation,
+  maxOperation,
+} from './operation_definitions/metrics';
+import { dateHistogramOperation } from './operation_definitions/date_histogram';
+import { countOperation } from './operation_definitions/count';
+
+type PossibleOperationDefinitions<
+  U extends IndexPatternColumn = IndexPatternColumn
+> = U extends IndexPatternColumn ? OperationDefinition<U> : never;
+
+type PossibleOperationDefinitionMapEntyries<
+  U extends PossibleOperationDefinitions = PossibleOperationDefinitions
+> = U extends PossibleOperationDefinitions ? { [K in U['type']]: U } : never;
+
+type UnionToIntersection<U> = (U extends U ? (k: U) => void : never) extends ((k: infer I) => void)
+  ? I
+  : never;
+
+export type AllOperationDefinitions = UnionToIntersection<PossibleOperationDefinitionMapEntyries>;
+
+export const operationDefinitionMap: AllOperationDefinitions = {
```

Comment:
I'm not quite following all the logic here, but it seems like you commented later that this object type is not strong enough. Was there an issue that makes it not possible to simplify like `type AllOperationDefinitions = Record<OperationType, OperationEntry>`?

Reply:
I added a comment for that - this typing makes sure that
1. each defined operation type also has an operation definition in this map
2. the key of a definition is the name of the operation type

This makes sure on a type level that everything around a newly added operation is actually provided.

The comment later is there because typescript can't infer that correctly in any case, even though it's guaranteed by this type that this is always consistent. `Record<OperationType, OperationEntry>` doesn't ensure this relationship. No strong opinion on that though, maybe I went a bit overboard with these types :)